### ISSUE TITLE
Where `application/x-www-form-urlencoded` and `multipart/form-data`?

### ISSUE THREAD

**Issue by AlexeyMatskevich** [hanami/controller#439]
I'm trying to contribute to the rspec-openapi library and I'm using hanami 2 in production. In [rspec-openapi](https://github.com/exoego/rspec-openapi/tree/master/spec/rails), there is a dummy test application on rails that I am trying to re-implementon hanami.

In particular, the original application in some cases transmits data in the following formats: `application/x-www-form-urlencoded` and `multipart/form-data` when upload images.

I'm reading this article https://guides.hanamirb.org/v2.1/actions/formats-and-mime-types/ and I'm following this link [over 50 of the most common MIME types](https://github.com/hanami/controller/blob/dc5bb2a1db48b0ccf3faf52aac20eaef0fd135a3/lib/hanami/action/mime.rb#L15-L69) and I can't find the MIME-type I need in the list. 

By default I'm using form :json, so I'm trying to get around the problem manually:

```ruby
module HanamiTest

  class App < Hanami::App
    config.actions.formats.add(:form, "application/x-www-form-urlencoded")
    config.actions.formats.add(:multipart, "multipart/form-data")
    config.actions.format :json, :form, :multipart
  end
end
```
But I only succeed with form data. 

I researched `Hanami::Action::Mime.accepted_mime_type? 
```ruby
        def accepted_mime_type?(mime_type, config)
          accepted_mime_types(config).any? { |accepted_mime_type|
            ::Rack::Mime.match?(mime_type, accepted_mime_type)
          }
        end
```
i get
```ruby
irb(Hanami::Action::Mime):001:0> mime_type
=> "multipart/form-data; boundary=----------XnJLe9ZIbbGUYtzPQJ16u1"
irb(Hanami::Action::Mime):002:0> accepted_mime_types(config)
=> ["multipart/form-data"]
```

I also found this code for [FormParser](https://github.com/hanami/router/blob/97f75b8529574bd4ff23165460e82a6587bc323c/lib/hanami/middleware/body_parser/form_parser.rb#L11) but couldn't find any documentation for it and have no idea how to use it.

Any help would be appreciated.

**Comment by AlexeyMatskevich** [comment]
I found a workaround, remove format everywhere above inheritance and from app.rb
Action
```
module HanamiTest
  class Action < Hanami::Action
    format :json - remove
  end
end
```
App.rb
```
module HanamiTest
  class App < Hanami::App
    config.actions.format :json - remove
  end
end
```

**Comment by cllns** [comment]
@AlexeyMatskevich Thanks for reporting this issue. I've opened a PR adding `:form` and `:multipart` as built-in formats, which will be very helpful.

I think your "workaround" above is actually the solution to this problem. The config formats are the highest level and you were manually overriding them lower down the stack in your actions, so they weren't being respected. Does that sound correct?

**Comment by AlexeyMatskevich** [comment]
@cllns Yeah, that sounds about right! For me, this problem comes down to this example - “If I have a basic API application and I define format :json in the Action class - then I can't override the format in subclasses when I need to use multipart or form mimetype in one or two places, simply because that mimetype doesn't exist”.

**Comment by timriley** [comment]
This should now be fixed by https://github.com/hanami/hanami-controller/pull/485.

As before, if your actions do not have any format configuration, then requests with `Content-Typoe` of `"application/x-www-form-urlencoded"` or `"multipart/form-data"` are accepted.

This is now _also_ the case if your actions _do_ have format config. So long as they accept the `:html` format, the above `Content-Types` will _still_ be permitted.

So in your examples, this would be a matter of doing:

```ruby
config.formats.accept :html, :json
```

The one missing piece is making sure you have the body parser enabled for `"multipart/form-data"` requests. I'm going to address this in a separate PR, and will close this issue after that is done :)

**Comment by timriley** [comment]
And here's the PR where we parse `multipart/form-data` bodies by default, no user config required: https://github.com/hanami/hanami/pull/1549

Once this merges, I think we can consider this issue resolved.

**Comment by timriley** [comment]
This is all merged now :) It will be released with Hanami 2.3 shortly.
